# Figures

![Phylogeny of the 138 genomes of this study that had whole-genome sequence data available, based on SNPs found in regions present in at least 70 % of the genomes.](./images/snp_tree.png)

<!-- 
![The effect of kmer sizes 3 - 31 and kmer frequency cutoff values of 1 - 15 in the correct prediction of host-source among _E. coli_ O157:H7 using WGS from the Lupolova et al. (2016) dataset [@Lupolova2016].](./images/KmerLength.pdf){#fig:kmer}

![The effect of various feature selection methods and the amount of features removed in the correct prediction of host-source amoung _E. coli_ O157:H7 using WGS from the Lupolova et al. (2016) dataset [@Lupolova2016].](./images/FeatureSelection.pdf){#fig:featureselection}

![The accuracy of Neural Network, Random Forest, and Support Vector machine models in predicting serotype from whole-genome sequence and omnilog microarray data, for all serotypes with five or more bacterial isolates](./images/Serotype.pdf){#fig:serotype}

![The accuracy of Neural Network, Random Forest, and Support Vector machine models in predicting O-group from whole-genome sequence and omnilog microarray data, for all O-groups with five or more bacterial isolates](./images/Otype.pdf){#fig:otype}

![The accuracy of Neural Network, Random Forest, and Support Vector machine models in predicting H-group from whole-genome sequence and omnilog microarray data, for all H-groups with five or more bacterial isolates](./images/Htype.pdf){#fig:htype}

![Multiclass Figure Title](./images/Multiclass.pdf){#fig:multiclass}

![Host Figure Title](./images/Host.pdf){#fig:hosts} -->

\newpage